### ISSUE TITLE
abstract.py's writeSequence implementation accepts unicode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11526]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1510|
|Type|defect|
|Created|2006-02-27 16:56:03Z|

It used to reject it, but I guess when it got changed to be more efficient that behavior got lost.

Here's a trivial patch that puts this behavior back:

```
Index: twisted/internet/abstract.py
===================================================================
--- twisted/internet/abstract.py        (revision 16071)
+++ twisted/internet/abstract.py        (working copy)
@@ -177,6 +177,8 @@
             self.startWriting()
 
     def writeSequence(self, iovec):
+        if unicode in map(type, iovec):
+            raise TypeError("iovec must not contain unicode data")
         if not self.connected or not iovec or self._writeDisconnected:
             return
         self._tempDataBuffer.extend(iovec)
```

Although there should be a unit test for this, too.

Attachments:

* [patch_1510.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9a6/9a6aa8b8b6919d3b97d0f40c9eda85f5523c7dd2/7ddcdcc90b328eac7ae41bd28ff645e07d7150bb.patch) (1836 bytes) - added by moijes12 on 2012-02-03 11:35:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1510 1510
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__ 
time__1141059363000000 1141059363000000
changetime__1328317084000000 1328317084000000
version__ 
owner__ 
cc__exarkun cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Maybe twisted can just install a half-asscii codec that only allows conversion from ascii-str->unicode but not unicode->ascii-str as the default.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @therve|
|-|-|

I'll do that once [#10493](https://github.com/twisted/twisted/issues/10493) is in.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

Thanks.  This was already fixed in r33003, though, as part of [#3896](https://github.com/twisted/twisted/issues/3896), which was sort of a duplicate of this ticket.  The implementation in trunk now is fine.
